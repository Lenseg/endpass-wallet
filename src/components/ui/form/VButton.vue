<template>
  <div class="field">
    <div class="control">
      <button :id="id" @click="$emit('click',$event)"
         :disabled="!form.isFormValid || disabled"
         class="button"
         :type="type"
         :class="[...classes, {'is-loading' : loading }]">
        <slot />
      </button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'v-button',
  inject: {
    $validator: '$validator',
    form: {
      default: () => ({
        isFormValid: true,
      }),
    },
  },
  props: {
    id: {
      type: String,
      default: null,
    },
    className: {
      type: String,
      default: '',
    },
    type: {
      type: String,
      default: 'submit'
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    loading: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classes() {
      return this.className.split(' ');
    },
  }
};
</script>

<style lang="scss">
</style>
