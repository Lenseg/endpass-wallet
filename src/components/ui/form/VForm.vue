<template>
  <form :id="id">
    <slot />
  </form>
</template>

<script>
export default {
  name: 'v-form',
  props: {
    id: {
      type: String,
      default: '',
    },
  },
};
</script>

<style lang="scss">
</style>
