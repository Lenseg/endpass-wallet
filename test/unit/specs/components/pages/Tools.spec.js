import { shallow, createLocalVue } from '@vue/test-utils';
import Notifications from 'vue-notification'
import VForm from '@/components/ui/form/VForm.vue';
import VButton from '@/components/ui/form/VButton.vue';
import ToolsPage from '@/components/pages/Tools.vue';

describe('Tools page', () => {
  const signedMessage = {};
  let wrapper;

  beforeEach(() => {
    const localVue = createLocalVue();
    const $store = {
      state: {
        accounts: {
          wallet: {
            getPrivateKey: jest.fn(() => 'private key')
          }
        },
        web3: {
          web3: {
            eth: {
              accounts: {
                sign: jest.fn(() => signedMessage)
              }
            }
          }
        }
      }
    };

    localVue.use(Notifications);

    wrapper = shallow(ToolsPage, {
      localVue,
      mocks: { $store },
      stubs: {
        'v-form': VForm,
        'v-button': VButton
      }
    });
  });

  describe('render', () => {
    it('should render the initial state of the page', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should enable "Sign message" button', () => {
      wrapper.setData({
        signingMessage: {
          message: 'message'
        }
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should show signed message', () => {
      const signedMessage = { message: 'message' };

      wrapper.setData({
        signingMessage: { signedMessage }
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('togglePasswordModal', () => {
      it('should toggle password modal', () => {
        const { vm } = wrapper;
        let isPasswordModal = vm.isPasswordModal;

        vm.togglePasswordModal();
        expect(vm.isPasswordModal).toBe(!isPasswordModal);

        isPasswordModal = vm.isPasswordModal;
        vm.togglePasswordModal();
        expect(vm.isPasswordModal).toBe(!isPasswordModal);
      });
    });

    describe('signMessage', () => {
      const password = 'password';

      beforeEach(() => {
        spyOn(wrapper.vm, '$notify');
      });

      it('should sign message', () => {
        const { vm } = wrapper;

        vm.signMessage(password);

        expect(vm.signingMessage.signedMessage).toEqual(signedMessage);
        expect(vm.$notify).not.toHaveBeenCalled();
      });

      it('should not sign message', () => {
        const { vm } = wrapper;

        vm.$store.state.web3.web3.eth.accounts.sign = jest.fn(() => {
          throw new Error();
        });
        vm.signMessage(password);

        expect(vm.signingMessage.signedMessage).toBeNull();
        expect(vm.$notify).toHaveBeenCalledTimes(1);
        expect(vm.$notify).toHaveBeenCalledWith({
          title: 'Error signing message',
          text: 'An error occurred while signing the message. Please try again.',
          type: 'is-danger',
        });
      });
    });
  });
});
