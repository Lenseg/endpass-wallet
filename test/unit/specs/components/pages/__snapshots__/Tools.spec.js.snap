// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tools page render should enable "Sign message" button 1`] = `
<div
  class="app-page tools-page"
>
  <div
    class="section"
  >
    <div
      class="container is-narrow"
    >
      <div
        class="card app-card main-app-card"
      >
        <div
          class="card-header"
        >
          <h1
            class="card-header-title"
          >
            Tools
          </h1>
        </div>
         
        <div
          class="card-content"
        >
          <form
            id="signing-message-form"
          >
            <div
              class="field"
            >
              <label
                class="label"
              >
                Message
              </label>
               
              <div
                class="control"
              >
                <textarea
                  class="textarea"
                  placeholder="This is a message that you are signing to prove that you own the address you say you own."
                />
              </div>
            </div>
             
            <div
              class="field"
            >
              <div
                class="control"
              >
                <a
                  class="button is-primary is-medium"
                >
                  Sign message
                </a>
              </div>
            </div>
             
            <!---->
          </form>
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`Tools page render should render the initial state of the page 1`] = `
<div
  class="app-page tools-page"
>
  <div
    class="section"
  >
    <div
      class="container is-narrow"
    >
      <div
        class="card app-card main-app-card"
      >
        <div
          class="card-header"
        >
          <h1
            class="card-header-title"
          >
            Tools
          </h1>
        </div>
         
        <div
          class="card-content"
        >
          <form
            id="signing-message-form"
          >
            <div
              class="field"
            >
              <label
                class="label"
              >
                Message
              </label>
               
              <div
                class="control"
              >
                <textarea
                  class="textarea"
                  placeholder="This is a message that you are signing to prove that you own the address you say you own."
                />
              </div>
            </div>
             
            <div
              class="field"
            >
              <div
                class="control"
              >
                <a
                  class="button is-primary is-medium"
                  disabled="disabled"
                >
                  Sign message
                </a>
              </div>
            </div>
             
            <!---->
          </form>
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`Tools page render should show signed message 1`] = `
<div
  class="app-page tools-page"
>
  <div
    class="section"
  >
    <div
      class="container is-narrow"
    >
      <div
        class="card app-card main-app-card"
      >
        <div
          class="card-header"
        >
          <h1
            class="card-header-title"
          >
            Tools
          </h1>
        </div>
         
        <div
          class="card-content"
        >
          <form
            id="signing-message-form"
          >
            <div
              class="field"
            >
              <label
                class="label"
              >
                Message
              </label>
               
              <div
                class="control"
              >
                <textarea
                  class="textarea"
                  placeholder="This is a message that you are signing to prove that you own the address you say you own."
                />
              </div>
            </div>
             
            <div
              class="field"
            >
              <div
                class="control"
              >
                <a
                  class="button is-primary is-medium"
                  disabled="disabled"
                >
                  Sign message
                </a>
              </div>
            </div>
             
            <div
              class="field"
            >
              <label
                class="label"
              >
                Signature
              </label>
               
              <div
                class="control"
              >
                <textarea
                  class="textarea"
                  disabled="disabled"
                />
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
